[upgrade] Use FFT, various fixups, add more settings

kujub opened the pull request:
Hi all!
First of all this will make it actually work like a spectrum analyzer.
Note: Changes for GL only (no directx)

Additional  fixes and improvements:
- cleanup code
- fix swapped height-speed setting values and enable one more option to disable delaying
- Render: fix wireframe mode and performance
- perspective: honor display and pixel aspects to fix excessive distortion, add some settings
- ... (see below)

AlwinEsch commented on visualization.spectrum/resources/settings.xml:
```diff
@@ -68,6 +69,46 @@
+            <formatlabel>30024</formatlabel>
+            <popup>false</popup>
+          </control>
+        </setting>
```
As this three new settings only works with GL must you add the:
```xml
          <dependencies>
            <dependency type="visible">
              <or>
                <condition on="property" name="IsDefined">HAS_GL</condition>
                <condition on="property" name="IsDefined">HAS_GLES</condition>
              </or>
            </dependency>
          </dependencies>
```
to prevent show on Windows and his DX.

kujub replied:
> As only part on view where seems to improved, is when no sound comes anymore, There stays one row still a bit up:

This was never seen here. If you paused, it is maybe normal behavior because shifting rows is paused too then.

> As this three new settings only works with GL must you add the:
>           <dependencies>
> to prevent show on Windows and his DX.

Done

AlwinEsch commented on visualization.spectrum/resources/language/resource.language.en_us/strings.po:
```diff
@@ -95,3 +95,23 @@ msgstr "%i°"
+
+msgctxt "#30024"
+msgid "%i%%"
+msgstr "%i%%"
```
With formal labels was something changed inside Kodi and with this style now no more supported (on Nexus) and need:
```po
msgid "{0:d}%"
```

kujub replied:
Done

AlwinEsch commented on visualization.spectrum/resources/language/resource.language.en_us/strings.po:
```diff
@@ -95,3 +95,23 @@ msgstr "%i°"
+
+msgctxt "#30022"
+msgid "%+i%%"
+msgstr "%+i%%"
```
What makes the `+` before the `i`, not know as this, makes it only then number if higher as 0?

kujub replied:
Was just kinda cosmetics because actually "0%" is 100% (default) and "+10%" is 110% and so on.
Removed now as i don't know how to do this with the new format.

AlwinEsch commented on visualization.spectrum/resources/language/resource.language.de_de/strings.po:
```diff
@@ -95,3 +95,23 @@ msgstr "%i°"
+
+msgctxt "#30024"
+msgid "%i%%"
+msgstr ""
```
As with use now the translation system by Weblate, would it better to leave de_de and the en_us unchanged.

kujub replied:
Done (reverted)

AlwinEsch reviewed (changes requested):
Becomed time to look. Your changes looks good and makes it better, only few parts need changed, see notes on related places.

As only part on view where seems to improved, is when no sound comes anymore, There stays one row still a bit up:
![Bildschirmfoto vom 2021-07-17 12-13-43](https://user-images.githubusercontent.com/6879739/126034289-b6f44597-7303-4553-8f37-0bb99bb860e0.png)

AlwinEsch commented:
As Note about Windows, I think also to kick the Windows DX support complete and to make via the "Angle" as GLES with it.

There then also the Question has you tested with GLES and worked correct?
And thanks a lot to look for improvements here, your help is welcome :+1:

AlwinEsch commented:
Have changed the target branch to "Nexus", after them here final can we create a backport about for Kodi Matrix.

kujub commented:
> There then also the Question has you tested with GLES and worked correct?

Tested GL only (on Fedora 33 x86-64). It seems kodi is built with support for GL and GLES, but i don't know how to switch.

kujub commented:
> > There then also the Question has you tested with GLES and worked correct?
> 
> Tested GL only (on Fedora 33 x86-64). It seems kodi is built with support for GL and GLES, but i don't know how to switch.

Update: Rebuilt kodi 19.1 with -DAPP_RENDER_SYSTEM=gles. Visualization failed with:
`ERROR <general>: AddOnLog: visualization.spectrum: CShaderProgram::CompileAndLink: error: vertex shader lacks `main'`
[kodi.log](https://github.com/xbmc/visualization.spectrum/files/6835697/kodi.log)

kujub commented:
> > There then also the Question has you tested with GLES and worked correct?
 > 
 > 
 > Tested GL only (on Fedora 33 x86-64). It seems kodi is built with support for GL and GLES, but i don't know how to switch.
 
Update: Rebuilt kodi 19.1 with -DAPP_RENDER_SYSTEM=gles. Then rebuilt visualization. Appears to work.
[kodi.log](https://github.com/xbmc/visualization.spectrum/files/6835713/kodi.log)

AlwinEsch commented:
Has also tested now with GLES and works, with my reported bars seems related to the sound format on played "m4a" files, by other formats it was not present.

From my side is the request now OK and can in. Only if possible can you cleanup the commits a bit, so that the last correction commits would no longer be necessary.

With your request works now with his bars much more correct!

kujub commented:
> Has also tested now with GLES and works, with my reported bars seems related to the sound format on played "m4a" files, by other formats it was not present.

OK. I only played mp3 and flac.

> From my side is the request now OK and can in. Only if possible can you cleanup the commits a bit, so that the last correction commits would no longer be necessary.

Hmm... git rebase; git push -f appears a bit dangerous to me. Github recommends applying suggestion blocks, but is this still possible now? Wouldn't the correction commits cause conflicts preventing (the same) changes on previous commits? Do I get a fork of a fork then? [confused]

kujub commented:
> From my side is the request now OK and can in. Only if possible can you cleanup the commits a bit, so that the last correction commits would no longer be necessary.

Squashed all fixup commits also dropping some trailing white-space.

kujub commented:
Note: Actually newly added: SetSetting cleanup, Rotation speed setting. Perspective settings just rebased for commit cleanup.

AlwinEsch commented:
About this request must be also checked by other Team Member's, like to use FFT, thats why for me OK. Only not complete sure about human feeling on his current displayed bars, as them mostly all complete up and slightly move then up/down to show.

Here as images:
- WIth this request on 1000Hz Sinus tone
![Bildschirmfoto von 2021-09-04 16-40-38](https://user-images.githubusercontent.com/6879739/132099607-fadbb44b-ead0-4aa7-ab60-71632d945c4e.png)
- Here before on 1000Hz Sinus tone
![Bildschirmfoto von 2021-09-04 16-42-57](https://user-images.githubusercontent.com/6879739/132099612-e803b1e4-ba0d-46cd-b87d-1c9f58fe8e2d.png)
- Here with new on normal played audio:
![Bildschirmfoto von 2021-09-04 16-51-42](https://user-images.githubusercontent.com/6879739/132099586-d56490ef-2338-459f-8d11-c96bdee32ee2.png)
- And here how on version before:
![Bildschirmfoto von 2021-09-04 16-53-09](https://user-images.githubusercontent.com/6879739/132099595-043c1430-0a8c-4006-9256-7d0b8962515e.png)

kujub commented:
> Only not complete sure about human feeling on his current displayed bars, as them mostly all complete up and slightly move then up/down to show.

a) The bar heights are currently scaled to the full dynamic range of CDDA (96 dB ) which is almost never really used. We could add one more setting to allow smaller ranges too.

b) About sine-waves: Ideally there should be only one row of bars up, but only if one of the FFT-frequencies is **exactly** matched. The wave must fit exactly into the probing window used so that the beginning matches the end to really have a sine-wave detected. The only perfect frequency I found so far is 11025 Hz. Other visualization addons seem to implement their own FFT-code, so I guess kodi's built in one isn't absolutely perfect.

AlwinEsch commented:
Thanks for the update will look next time.

With the bars seen, by work on another addon, a good place how I expect with FFTW. Here by ShaderToy https://www.shadertoy.com/view/4ljGD1 it shows below on "iChannel0" some bars where nice go up by related sound frequency.

kujub commented:
Note:

1. Rebased once more to squash two incomplete commits into one (draw_bars) [d0c68bd]
2. Additionally forked Nexus and Omega branches and merged in changes from 'Matrix' (topic branch). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/Omega ?

kujub commented:
1. Added new commit for API changes b7aeb17
2. Rebased once more to squash two incomplete commits into one (draw_bars) d0c68bd
3. Extra: Also forked Nexus and Omega branches and merged in changes from 'Matrix' (topic branch). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/Omega ?

edit: mention new commit added

kujub commented:
1. Added new commit for API changes
2. Rebase: a) solve merge conflicts; b) squash two incomplete commits [draw_bars d0c68bd]; c) fixup new API commit (kick out translations)
3. Extra: Also forked Nexus and Omega branches and merged in changes from 'Matrix' (topic branch). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/Omega ?

kujub commented:
Updates:
- Squashed two incomplete commits [draw_bars d0c68bd]
- Added new commit for FFT (API changes)
Extra: Also forked Nexus and Omega branches and merged in changes from this topic branch ('Matrix'). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/Omega ?

kujub commented:
Updates:

- Squashed two incomplete commits [draw_bars d0c68bd]
- Added new commit for FFT (API changes)
- Fixed the broken Hann function
@AlwinEsch : Now you should really get only one bar up with 1000 Hz sine (or other high frequencies).

Note: Also forked Nexus and Omega branches and merged in changes from this topic branch ('Matrix'). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/Omega ?

kujub commented:
Updates:
- Squashed two incomplete commits [draw_bars d0c68bd]
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function
@AlwinEsch : Now you should really get only one bar up with 1000 Hz sine (or other high frequencies).

Note: Also forked Nexus and Omega branches and merged in changes from this topic branch ('Matrix'). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/Omega ?

kujub commented:
Note:
New commit 'Add Bands setting' removed again because of merge conflicts.
Created new topic branch (for next PR): https://github.com/kujub/visualization.spectrum/tree/Omega-bands

kujub committed:
GL: More color modes, add zero height, points mode: single points
New default color mode (beside of legacy mode): Up to two bars are stacked with blue->green, green->red gradients from bottom to some constant heights. So top colors depend on relative bar heights.

New monochrome color modes: Monochrome gradient from dark to bright amber/blue/green.

New bar height setting 'Flat': Allow clean 2D views with changing bar top colors.

Points mode: Only draw one point at top mid.

kujub commented:
Updates:
- Squashed two incomplete commits [draw_bars d0c68bd]
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)

@AlwinEsch : Now you should really get only one bar up with 1000 Hz sine (or other high frequencies).

Note: Also forked Omega branch into 'pr-1' and merged in changes from this topic branch ('Matrix'). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/pr-1 ?

Additional commits in branch 'pr-2' https://github.com/kujub/visualization.spectrum/tree/pr-2 (maybe next PR):
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color mode setting and gradient mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)

kujub committed:
GL: Add smooth animation and take account of fps
Implement smooth animation for backwards shift.
Take account of fps for bar height animation speed.

kujub commented:
Updates:
- Squashed two incomplete commits [draw_bars d0c68bd]
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)

@AlwinEsch : Now you should really get only one bar up with 1000 Hz sine (or other high frequencies).

Note: Also forked Omega branch into 'pr-1' and merged in changes from this topic branch ('Matrix'). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/pr-1 ?

Additional commits in branch 'pr-2' https://github.com/kujub/visualization.spectrum/tree/pr-2 (maybe next PR):
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Add Bar width setting
- Implement Smooth animation (and related On/Off setting)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see below)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)

@AlwinEsch : Now you should really get only one bar up with 1000 Hz sine (or other high frequencies).

Note: Also forked Omega branch into 'pr-1' and merged in changes from this topic branch ('Matrix'). Maybe better pull from https://github.com/kujub/visualization.spectrum/tree/pr-1 ?

Additional commits in branch 'pr-2' https://github.com/kujub/visualization.spectrum/tree/pr-2 (maybe next PR):
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Upgrade bar height settings to allow a cleaner 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and cleanup code (this is a big one)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Added Additional commits:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Upgrade bar height settings to allow a cleaner 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and cleanup code (this is a big one)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Added Additional commits:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Upgrade bar height settings to allow a cleaner 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and cleanup code (this is a big one)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Added Additional commits:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Upgrade bar height settings to allow a cleaner 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and cleanup code (this is a big one)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Upgrade bar height settings to allow a cleaner 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and cleanup code (this is a big one)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Made a new commit for FFT (API changes)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color gradient setting and 'Bar' mode (blue-green-red from bottom of bar to some constant height, so top colors depend on relative bar heights)
- Upgrade bar height settings to allow a cleaner 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and cleanup code (this is a big one)

kujub committed:
Prepare new screenshots (TODO: enable after DX->GL changeover)

kujub committed:
GL: Add bar width setting and fix field width

kujub committed:
GL: Add Bands setting and default to 32
Allow up to 128 x 128 bars.

kujub committed:
GL: Add presets and reverse rotation setting
Allow to start/stop/revert rotation seamlessly at any point by switching back and forth on a set of presets with normal/zero/reverse rotation settings by using hotkeys or to change any other settings parameters on the fly as long as the number of bands remains equal.
A settings category called 'Presets template' allows easy setup and change of a bunch of similar presets.

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Squashed some early code cleanup commits
- Made a new commit for FFT (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color mode setting and default to new 'Bar' mode (Up to two bars are stacked with blue>green, green>red gradients from bottom to some constant heights. So top colors depend on relative bar heights)
- Upgrade bar height settings to allow a clean 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets, presets template and reverse rotation setting (this is a big one)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Squashed some early code cleanup commits
- Made a new commit for FFT (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add Color mode setting and default to new 'Bar' mode (Up to two bars are stacked with blue>green, green>red gradients from bottom to some constant heights. So top colors depend on relative bar heights)
- Upgrade bar height settings to allow a clean 2D bird's-eye view (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets, presets template and reverse rotation setting (this is a big one)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Squashed some early code cleanup commits
- Made a new commit for FFT (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets, presets template and reverse rotation setting (this is a big one)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Squashed some early code cleanup commits
- Made a new commit for FFT (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting, cleanup SetSetting code (this is a big one)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Squashed some early code cleanup commits
- Made a new commit for FFT (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting, cleanup SetSetting code (this is a big one)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- Squashed some early code cleanup commits
- Made a new commit for FFT (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting, cleanup SetSetting code (this is a big one)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed into one commit)
- Made a new commit for FFT and perspective (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed into one commit)
- Made a new commit for FFT and perspective (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed into one commit)
- Made a new commit for FFT and perspective (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed into one commit)
- Made a new commit for FFT and perspective (API changes / add kissfft)
- Fixed the broken Hann function (!)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (no changes to directx)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed into one commit)
- Made a new commit for FFT and perspective (API changes / add kissfft - see commit msg)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add Bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to 'Bar', upgrade bar height settings (see commit msg)
- Implement Smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed)
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed)
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed)
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed)
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed)
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Squashed two incomplete commits into GL: draw_bars ...
- More code cleanup (squashed)
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub committed:
GL: Render: Fix wireframe mode and performance
Wireframe: Show all edges of bars/cuboids, but w/o confusing diagonal lines.
Performance: Don't call glDrawArrays separately for each single bar. Enable face-culling.

kujub committed:
GL: Cleanup code a bit

kujub committed:
GL: Add no delay height speed setting

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two incomplete commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub committed:
GL: Render: Fix perspective and add some settings
Take account of display and pixel aspects and allow to change field size, tilt and base (bottom of bars).
With zero tilt and negative base, perspective distortion can be entirely eliminated and all bars stand vertical then.
With 90° tilt a funny bird's-eye view can be achieved.
Additionally the size of the x*z-field can be changed. Smaller is useful for bird's-eye view and bigger for zero tilt with rotation fixed at 0° or 90° etc.
Rotation speed setting is added too.

kujub committed:
GL: Add kissfft and use dB scaling
Really implement spectrum analyser functionality
A new setting allows one to set the dynamic range displayed to whatever is used in practice (up to 144 dB for full 24 bit audio)

Note: A new class KFFTR was derived from MRFFT used in other visualizations:
* discard DC bin
* fix hardcoded stereo (just sum up all channels)
* fix broken Hann window function (!)

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two incomplete commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two incomplete commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two incomplete commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and default to bar gradient, add flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- Add new color modes and flat bar height setting
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- AddBars: new color modes, zero bar height, single points at top mid (if points mode), fix performance once more
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- New color modes, zero bar height, single points at top mid (if points mode), fix render performance once more
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Reword: added GL: prefix (because no changes for DX)
- Way more code cleanup
- Squashed two half way commits into new commit GL: Render: ...
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- More color modes, zero bar height, single points at top mid (if points mode)
- Implement smooth animation (and setting to disable)
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- Way more code cleanup
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- More color modes, zero bar height, single points at top mid (if points mode)
- Implement smooth animation
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on recent Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- More code cleanup
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- More color modes, zero bar height, single points at top mid (if points mode)
- Smooth animation
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots

kujub commented:
Updates:
- Rebased on recent Nexus (for Omega see https://github.com/kujub/visualization.spectrum/tree/pr-2)
- More cleanup
- New commit for FFT and perspective (API changes / add kissfft)
@AlwinEsch : With this you should really get only one bar up now with 1000 Hz sine (or other high frequencies).

Upgrades:
- Add bands setting (allow up to 128 x 128 bars)
- More color modes, zero bar height, single points at top mid (if points mode)
- Smooth animation
- Add presets and reverse rotation setting (see commit msg)
- Prepare new screenshots
